File: plugins/local_time/android/module/src/main/java/net/chentong/leaf/android/plugin/localtime/LocalTimePlugin.java
package net.chentong.leaf.android.plugin.localtime;

import net.chentong.leaf.android.plugin.LeafPlugin;

import org.json.JSONObject;

import java.util.TimeZone;

public class LocalTimePlugin implements LeafPlugin {
    @Override
    public String pluginName() {
        return "LocalTimePlugin";
    }

    @Override
    public boolean canHandle(String method) {
        return "local_time.get_timezone".equals(method);
    }

    @Override
    public void onMethodCall(LeafMethodCall call, Result result) {
        if (!"local_time.get_timezone".equals(call.getMethod())) {
            result.error(call.getRequestId(), -404, "method_not_implemented", false);
            return;
        }

        try {
            final String timezone = resolveTimezone();
            if (timezone.isEmpty()) {
                result.error(call.getRequestId(), -2, "timezone_empty", false);
                return;
            }

            JSONObject out = new JSONObject();
            out.put("timezone", timezone);
            result.success(call.getRequestId(), out.toString());
        } catch (Throwable throwable) {
            result.error(call.getRequestId(), -500, "get_timezone_failed", false);
        }
    }

    private String resolveTimezone() {
        TimeZone timezone = TimeZone.getDefault();
        if (timezone == null) {
            return "";
        }

        String id = timezone.getID();
        return id == null ? "" : id;
    }
}

File: plugins/local_time/ios/module/Classes/LocalTimePlugin.m
#import "LocalTimePlugin.h"

static const int32_t REQUEST_ERROR_INTERNAL = -500;
static const int32_t REQUEST_ERROR_METHOD_NOT_IMPLEMENTED = -404;

@implementation LocalTimePlugin

- (NSString *)pluginName {
    return @"LocalTimePlugin";
}

- (BOOL)canHandle:(NSString *)method {
    return [method isEqualToString:@"local_time.get_timezone"];
}

- (void)onMethodCall:(id<LeafMethodCall>)call result:(id<LeafResult>)result {
    NSString *method = call.method ?: @"";
    if ([method isEqualToString:@"local_time.get_timezone"]) {
        [self handleGetTimezone:call result:result];
        return;
    }
    [result error:call.requestId
             code:REQUEST_ERROR_METHOD_NOT_IMPLEMENTED
            error:@"method_not_implemented"
         canceled:NO];
}

- (void)handleGetTimezone:(id<LeafMethodCall>)call result:(id<LeafResult>)result {
    NSString *timezone = NSTimeZone.localTimeZone.name;
    if (timezone.length == 0) {
        [result error:call.requestId
                 code:REQUEST_ERROR_INTERNAL
                error:@"timezone_unavailable"
             canceled:NO];
        return;
    }
    NSDictionary *payload = @{@"timezone": timezone};
    NSString *json = [self jsonStringFromDictionary:payload];
    if (!json) {
        [result error:call.requestId
                 code:REQUEST_ERROR_INTERNAL
                error:@"serialize_result_failed"
             canceled:NO];
        return;
    }
    [result success:call.requestId data:json];
}

- (nullable NSString *)jsonStringFromDictionary:(NSDictionary *)dictionary {
    NSError *error = nil;
    NSData *data = [NSJSONSerialization dataWithJSONObject:dictionary options:0 error:&error];
    if (!data || error) {
        return nil;
    }
    return [[NSString alloc] initWithData:data encoding:NSUTF8StringEncoding];
}

@end
